src/ray/rpc/client_call.h

```diff
@@ -3,27 +3,109 @@
 
 #include <grpcpp/grpcpp.h>
 #include <boost/asio.hpp>
+#include <queue>
 
 #include "ray/common/grpc_util.h"
 #include "ray/common/status.h"
 
 namespace ray {
 namespace rpc {
 
+/// The type of the client call.
+enum class ClientCallType {
+  DEFAULT_ASYNC_CALL,
+  STREAM_ASYNC_CALL,
+};
+
+enum class ClientCallState {
+  CONNECT,
+  WRITING,
+  PENDING,
+  WRITES_DONE,
+};
+
+class ClientCallTag;
 /// Represents an outgoing gRPC request.
+/// It's the common parts of all types of client calls.
+/// The interfaces that should be implemented by both types of client calls are defined as
+/// pure virtual function. Otherwise we provide a default implementation to avoid unused
+/// definitions in the subclass.
 ///
 /// NOTE(hchen): Compared to `ClientCallImpl`, this abstract interface doesn't use
 /// template. This allows the users (e.g., `ClientCallMangager`) not having to use
 /// template as well.
 class ClientCall {
  public:
-  /// The callback to be called by `ClientCallManager` when the reply of this request is
-  /// received.
+  /// Interfaces for default async call or stream async call.
+
+  /// Callback which will be called once the client received the reply.
   virtual void OnReplyReceived() = 0;
-  /// Return status.
-  virtual ray::Status GetStatus() = 0;
+
+ public:
+  /// Interfaces only for stream async call.
+
+  /// Only one write may be outstanding at any given time. This means that
+  /// after calling `Write`, one must wait to receive a tag from the completion
+  /// queue before calling `Write` again.
+  /// Try to write a request into the rpc stream. If stream is not ready
+  /// to write, we will write it into a writer buffer first to avoid calling `Write`
+  /// twice, which will cause grpc crash.
+  ///
+  /// \param request The request message.
+  /// \return False if the stream is not ready and the buffer is full.
+  virtual bool WriteStream(const ::google::protobuf::Message &request) {}
```

Comment:
for these interfaces that are specific to one type of `ClientCall`, I think it would be much better to put them into derived classes, and do a downcast when necessary.